### ISSUE TITLE
Move to the official bazelbuild/rules_rust?

### ISSUE THREAD

**Issue by acmcarther** [google/cargo-raze#12]
Reraising @bspeice's issue from https://github.com/google/cargo-raze/pull/11

The reason that this isn't in rules_rust was that I wasn't sure that this was the right solution for everyone to pursue. Solutions to Cargo-Bazel build system integration seems to lie on a spectrum between "let Cargo make decisions" and "let Bazel make decisions", and this project is decidedly in the latter camp. 

Most folks I've talked to on both sides expect Cargo to play a much larger role than it plays in this implementation -- it just turned out in practice that it was easier for me to get code compiling by structuring the integration in the manner that I did. That said, I've been averse to imply that cargo-raze is the way an integration "must be". 

Its possible that an implementation similar to `gazelle` that treats BUILD files monorepo-wide as generated artifacts of Cargo.tomls would be more compatible with the ecosystem. I personally don't like the development experience that Cargo provides though, so this was as much an escape from that workflow as it was a compatiblity project.

---

Footnote: My issues are primarily:
- The somewhat artificial notion of a "project". 
- "Cargo.toml" files being mandatory per "project" (I can't fault the ecosystem here, its necessary in the project model)
- "src/" as a thing. If every piece of code in my project had to be under "src", i'd be wearing out my 's', 'r', and 'c' characters while navigating the filetree in my terminal.
- Cargo+Rust's co-opting of the directory structure. As a practice, I don't let rust_libraries bring in stuff from nested directories in the form
```rust
mod expected_lib;
// Automagically bring in expected_lib/mod.rs
```

**Comment by acmcarther** [comment]
For comparison, in one of my projects, I structure my code like this, without `src/` or Cargo.toml outside `//third_party/cargo` etc.

```
├── core
│   ├── alg
│   │   ├── BUILD
│   │   └── octree.rs
│   ├── BUILD
│   └── init.rs
├── experimental
│   ├── BUILD
│   ├── renderer
│   │   ├── galaxy_toy
│   │   │   ├── BUILD
│   │   │   ├── explosion.png
│   │   │   ├── galaxy_toy.rs
│   │   │   ├── star.frag
│   │   │   └── star.vert
│   │   └── planet
│   │       ├── BUILD
│   │       ├── planet.frag
│   │       ├── planet_renderer.rs
│   │       ├── planet.rs
│   │       └── planet.vert
├── game
│   ├── assets
│   │   ├── BUILD
│   │   ├── gen
│   │   │   ├── BUILD
│   │   │   └── icosphere.rs
│   │   └── geometry.rs
│   ├── client
│   │   ├── BUILD
│   │   └── renderer
│   │       ├── BUILD
│   │       ├── renderer.rs
│   │       ├── sdl2_vulkan_interop.rs
│   │       ├── vk_buffer_support.rs
│   │       ├── vk_descriptor_support.rs
│   │       ├── vk_device_support.rs
│   │       ├── vk_instance_support.rs
│   │       ├── vk_lite.rs
│   │       ├── vk_pipeline_support.rs
│   │       └── vk_swapchain_support.rs
```

**Comment by bspeice** [comment]
I guess the biggest question would be: what is the benefit of integrating this into rules_rust? From what I understand so far, the purpose of `raze` is to make the transition from Cargo to Bazel as easy as possible, without imposing any ideology on either side of Bazel's "just compile things" rules or Cargo's "handle everything" method.

Keeping this separate leaves Cargo free to be the default solution for Rust, while keeping rules_rust focused on "pure" compilation/testing (and implicitly a more free-form project). It should definitely be made clear in rules_rust that "if you want Cargo, we recommend cargo-raze," but otherwise raze seems to be a rather elegant solution to people wanting the best of both worlds.

**Comment by mfarrugi** [comment]
Random thoughts from interacting with tooling:

I have been hand writing a Cargo.toml for each of my BUILD files to get intellij-rust support, since my Cargo.toml generator is kind of fuzzy. We will definitely need a Cargo.toml to exist for the rest of the ecosystem to work well with bazel built code, but the duplication is not ideal (without saying anything about intellij-rust being its own sort of separate build system).

It might also be worth thinking about what having bazel read Cargo.toml's would like like.. but it feels like it gets messy fast. Metadeps sort of covers -sys crates and non-rust deps, but covering every other build.rs (or non cc_*/rust_* rule) use-case doesn't seem tractable. 

Do we want to be able to publish bazel built crates to cargo? At best we can have a build.rs that calls bazel.. but that feels backwards too.

**Comment by acmcarther** [comment]
@mfarrugi: What do the tools you use need from the Cargo.toml? I expect that RLS would like to be able to actually compile the code -- I don't think this is possible in general. You can trivially concoct a dependency graph that includes a genrule -- which would be non-trivial to port across the boundary without running `bazel` itself.

If tools aren't trying to run a full compilation, it'd probably be possible to get something that mostly works.

Or, it might be possible to use Bazel attributes to construct some sort of parallel file structure that performs the necessary substitutions such that RLS could actually compile it... though that sounds like a lot of effort.

**Comment by mfarrugi** [comment]
intellij-rust uses cargo metadata, and it's not using the rls at all. I've
already gotten most of the cases handle w/ the cargo.toml aspect but I'm
not confident on how it should be disemminated or that it isn't fragile.

**Comment by GregBowyer** [comment]
@mfarrugi I think maybe we should go the otherway and add the support to the bazel intellij plugin to have it configure rust support correctly.

**Comment by mfarrugi** [comment]
@GregBowyer I think that both plugins are fairly unstable, and getting them to play nice isn't something I would volunteer for.

@damienmg you mentioned recently you were working on this, how is it looking :)?

**Comment by damienmg** [comment]
Yes sorry I did not had time to work much on that recently. I will share my code as soon as I can go back to it but I can make bazel output Cargo.toml, metadata.json and rls-data.json files that should allow any of this tool to work fine with Bazel (still need to be tested fully).

RLS needs to have Cargo.toml or rls-data.json. It delegate to cargo or custom build system for the initial builds and have a heuristic to decide whether to call the build system or simply do an incremental build inside RLS. Although I don't think we can use that heuristic with a custom build system as there's seem to be no way to specify to RLS for the external build system how to pass arguments to the build system. Anyway this needs to be ironed out.

I would also hold back from importing into rules_rust until we can avoid generating build files.
